### PR TITLE
2.4.x - search by sku corrections

### PR DESCRIPTION
**Pull request by ericrisler**
This commit applies the patch and updates created by @southerncomputer in this Issue: https://github.com/Smile-SA/elasticsuite/issues/412.

Also found a bug in the setup:upgrade processes when InstallData for the virtual categories executed.

### CODE REVIEW

File: src/module-elasticsuite-virtual-category/Setup/InstallData.php [hunk 1]
```diff
@@ -45,11 +51,13 @@ class InstallData implements InstallDataInterface
...
      * @param EavSetupFactory             $eavSetupFactory             Eav setup factory.
      * @param VirtualCategorySetupFactory $virtualCategorySetupFactory Virtual Category setup factory.
+     * @param Config                  $eavConfig       EAV Config.
      */
-    public function __construct(EavSetupFactory $eavSetupFactory, VirtualCategorySetupFactory $virtualCategorySetupFactory)
+    public function __construct(EavSetupFactory $eavSetupFactory, VirtualCategorySetupFactory $virtualCategorySetupFactory, Config $eavConfig)
     {
         $this->eavSetupFactory      = $eavSetupFactory;
         $this->virtualCategorySetup = $virtualCategorySetupFactory->create();
+        $this->eavConfig       = $eavConfig;
```
Comment (Koc) [inline src/module-elasticsuite-virtual-category/Setup/InstallData.php]: This property seems unused

**Comment by houdang** [comment]
Any news on this merge?

**Comment by houdang** [comment]
@ericrisler Any plan when the build will be fixed?

**Comment by madonzy** [comment]
@Koc @ericrisler @southerncomputer does this pull request fixes error with partial searching of SKU? I mean if I have SKUs like RM-123, RM-32, RM-5345 and I'll type in the search field: "RM-" and find all products where SKU starts from "RM-" ?

**Comment by southerncomputer** [comment]
What I did eventually - was just throw the sku's into meta_keywords and make it searchable, but when importing the sku's i would break them up into their component pieces. I believe the problem with this solution is that the Analyzer_REFERENCE should break the sku into pieces and then the normal search should search against that field, but I don't believe the design of the indexer supports this method of breaking the sku into pieces then using the default search against that field

**Comment by ericrisler** [comment]
@madonzy - I don't think that this pull works on partial matching sku's with hypens. I just did a quick test on a site we use this for and it gave no results for a partial match. However, the autocomplete did pickup the correct partial matches. 

@southerncomputer - I didn't notice you gave feedback on improving the patch. I'll take a look at those those soon. Can you summarize what will the suggestions you make improve upon? Thanks.

**Comment by southerncomputer** [comment]
@ericrisler Since I could not figure out the edge n gram against SKU field, i just break the sku into chunks and throw it into searchable field meta_keywords. This allows the full-text search to work against the chunks of sku - basically what the EDGE NGRAM filter was trying to do to SKU field -
so sku HEW-ABCD1234 is broken on import to meta_keywords as:
hew-A,HEW-AB,HEW-ABC,HEW-ABCD,HEW-ABCD1,HEW-ABCD12,HEW-ABCD123, HEW-ABCD1234 plus the same for MFGPARTNO! It did not require all these modifications for search by sku and my import routines handle it with ease!

**Comment by afoucret** [comment]
Time for me to have an update on this one.

Search by SKUs is considered as fixed in branch 2.4.x of ElasticSuite :

* You can now find a product whose SKU is "AA-999df" even if you type "aa999 df". That a pretty big update.

* Search by prefix exists even if is not complete.
With the previous example the product can be found if you type "aa" or "aa999" not if you type "aa99". We don't support ngram and we will not in the future as explained bellow.

* Search by suffix is not supported out of the box and you will not find a product whose SKU is 001883 with the search 1883. It can be fixed with custom analyzer on the SKU field. In the B2C space it is a bad idea to enable search by suffix (exemple of 007 becoming 7) so we will not add it out-of-the box.

* We do not use ngram and are not huge fan of it. You can do it by yourself with custom analyzer but it will not be part of ElasticSuite. Here are some reasons :

    * Indexing performances are really bad when using ngram (and wasted space is too important).

    * It would undermine precision of the search as long we are not able to make the difference between search and index analyzer

* In version 2.5.x, I plan to introduce an new architecture for the autocomplete. It could allow a better product suggestion when typing SKUs.

**Comment by afoucret** [comment]
The proposed implementation is very different of what we plan to do with search by SKUs.
As a result I close this PR.
